languages/scala/reference/README.md

```diff
@@ -0,0 +1,193 @@
+# Scala reference
+
+## Concepts
+
+The Scala concept exercises are based on concepts. The list below contains the concepts that have been identified for the Scala language.
+
+### Language-unique
+
+- Pattern matching
+- Functional Programming
+- Immutability
+
+### Functional
+
+- [Functions][functions]
+  - [Pure functions][pure_functions]
+  - [Higher-order functions][higher_order_functions]
+  - [Anonymous functions][anonymous_functions]
+  - [Nested functions][nested_functions]
+  - [Function composition][function_composition]
+  - [Partial application][partial_application]
+  - [Recursion][recursion]
+    - Tail recursion
+  - Currying
+  - Implicit returns
+  - Type annotations
+- [Immutability][immutability]
+- [Pattern matching][pattern_matching]
+  - Discards
+  - Guard clauses
+- [Type inference][type_inference]
+  - Automatic generalisation
+- [Pipelines][pipelines]
+- [Expression-oriented (not statement oriented)][expression_oriented]
+
+### Object-oriented
+
+The core OO features an Scala developer should know about are:
+
+- [Classes][classes]
+- [case class][case class]
+- [Encapsulation][encapsulation]
+- [Objects][objects]
+- [Interfaces][interfaces]
+  - Object expressions
+- [Methods][methods]
+  - Constructors
+  - Apply
+  - Unapply
+  - Method arguments
+    - Named arguments
+    - [Optional arguments][optional_arguments]
+    - Out parameters
+- Fields
+- Properties
+- Type extensions
+- Implicit class
+
+While Scala has OO features, it is a "functional first" language. It is important to know how and when to apply the OO features, and which OO features to use.
+
+### General
+
+- Values
+  - Assignment
+- Scoping
+  - Namespaces
+  - Modules
+    - Extending modules
+  - Imports (`import`)
+  - Visibility (`public`, `private`, etc.)
+- Numbers
+  - Math operators
+  - Bitwise manipulation
+  - Arithmetic overflow
+- Comparison
+  - [Equality][equality] (`Equals`, `GetHashCode`)
+  - Ordering
+- Conditionals
+  - Boolean logic
+  - [Conditionals: if][conditionals]
+  - Conditionals: while
+- Enumeration
+  - [Enumeration: for loop][enumeration]
+- Collections
+  - [Generics][generics]
+  - Sequence/for expressions
+  - Iterators (yield)
+  - Collections: mapping
+  - Collections: filtering
+  - Collections: reducing
+  - Collections: ordering
+  - Collections: combining
+  - Ranges
+    - Slicing
+- Resources
+  - Resource passing (by reference/by value)
+- Conversions
+  - Implicit
+  - Explicit
+- Concurrency
+  - Concurrent collections `par`
+  - Future
+- Exceptions
+  - `Try`
+  - `Either`
+- Code quotations
+- Attributes
+- Nullability
+  - `Option`
+- String formatting
+  - Formatting types
+
+### Types
```

Comment:
Maybe we should add `Nothing`. Although I don't think that it's very important, I think it's interesting.